source/common/upstream/upstream_impl.cc

```diff
@@ -919,9 +919,16 @@ ClusterImplBase::ClusterImplBase(
 
   auto socket_matcher = std::make_unique<TransportSocketMatcherImpl>(
       cluster.transport_socket_matches(), factory_context, socket_factory, *stats_scope);
-  info_ = std::make_unique<ClusterInfoImpl>(cluster, factory_context.clusterManager().bindConfig(),
-                                            runtime, std::move(socket_matcher),
-                                            std::move(stats_scope), added_via_api, factory_context);
+  auto& dispatcher = factory_context.dispatcher();
+  info_ = std::shared_ptr<const ClusterInfoImpl>(
+      new ClusterInfoImpl(cluster, factory_context.clusterManager().bindConfig(), runtime,
+                          std::move(socket_matcher), std::move(stats_scope), added_via_api,
+                          factory_context),
+      [&dispatcher](const ClusterInfoImpl* self) {
+        ENVOY_LOG(debug, "Schedule destroy cluster info {}", self->name());
+        dispatcher.deleteInDispatcherThread(
+            std::unique_ptr<Event::DispatcherThreadDeletable>(const_cast<ClusterInfoImpl*>(self)));
```

Comment:
nit: deleteInDispatcherThread taking std::unique_ptr<const Event::DispatcherThreadDeletable> could save us a const_cast

Reply:
+1 would prefer to avoid a const_cast here.

Reply:
Yes.  Have the argument to dispatcher.deleteInDispatcherThread be:
``` std::unique_ptr<const Event::DispatcherThreadDeletable>&&```

Reply:
updated. Thanks!